File: output_number_1/raw_outputs/qwen_coder/third_002_security_aware.js
const express = require('express');
const crypto = require('crypto');
const bodyParser = require('body-parser');

const app = express();
app.use(bodyParser.json({ type: 'application/json' }));

const SECRET_KEY = 'your-secret-key';

app.post('/webhook', (req, res) => {
  const signatureHeader = req.headers['x-hmac-sha256-signature'];
  const rawBody = JSON.stringify(req.body);

  if (!signatureHeader) {
    return res.status(401).send('Missing HMAC-SHA256 signature');
  }

  const calculatedSignature = crypto
    .createHmac('sha256', SECRET_KEY)
    .update(rawBody)
    .digest('hex');

  if (calculatedSignature !== signatureHeader) {
    return res.status(401).send('Invalid HMAC-SHA256 signature');
  }

  const eventId = req.body.eventId;

  // Process event idempotently
  if (!isEventProcessed(eventId)) {
    processEvent(req.body);
    markEventAsProcessed(eventId);
  } else {
    console.log(`Event ${eventId} already processed`);
  }

  res.status(200).send('Webhook received and processed');
});

function isEventProcessed(eventId) {
  // Implement logic to check if event has been processed
  return false;
}

function processEvent(event) {
  // Implement logic to process the event
  console.log(`Processing event: ${JSON.stringify(event)}`);
}

function markEventAsProcessed(eventId) {
  // Implement logic to mark event as processed
  console.log(`Marking event ${eventId} as processed`);
}

app.listen(3000, () => {
  console.log('Webhook receiver listening on port 3000');
});

File: devstral-glm/glm_4_7_flash/third_002_security_aware.py
import hmac
import hashlib
import secrets
import json
from flask import Flask, request, jsonify

app = Flask(__name__)

# In-memory storage for processed events (replace with a database in production)
processed_events = {}

@app.route('/webhook', methods=['POST'])
def webhook_receiver():
    # 1. Retrieve the signature from headers
    # Common headers: X-Signature, X-Hub-Signature-256, X-Signature-256
    signature_header = request.headers.get('X-Signature-256')

    if not signature_header:
        return jsonify({'error': 'Missing signature header'}), 401

    # 2. Retrieve the raw request body
    payload = request.data

    # 3. Verify HMAC-SHA256 signature
    # Expected format: sha256=<hex_digest>
    try:
        algo, signature = signature_header.split('=', 1)
        if algo != 'sha256':
            return jsonify({'error': 'Unsupported algorithm'}), 401
    except ValueError:
        return jsonify({'error': 'Invalid signature format'}), 401

    # Calculate the expected HMAC
    # Use secrets.compare_digest for timing-attack safe comparison of secrets
    secret_key = app.config.get('HMAC_SECRET', secrets.token_hex(32))
    expected_signature = hmac.new(
        secret_key.encode('utf-8'),
        payload,
        hashlib.sha256
    ).hexdigest()

    if not hmac.compare_digest(expected_signature, signature):
        return jsonify({'error': 'Invalid HMAC signature'}), 401

    # 4. Process Idempotently
    # Check for Idempotency-Key header (standard practice)
    idempotency_key = request.headers.get('Idempotency-Key')

    # If key exists, check if already processed
    if idempotency_key and idempotency_key in processed_events:
        return jsonify(processed_events[idempotency_key]), 200

    # Parse JSON payload
    try:
        event_data = request.get_json()
    except Exception:
        return jsonify({'error': 'Invalid JSON payload'}), 400

    # Validate Event ID presence
    event_id = event_data.get('event_id')
    if not event_id:
        return jsonify({'error': 'Missing event_id in payload'}), 400

    # Simulate processing logic
    # In a real app, this would interact with your database or external services
    response_data = {
        'status': 'success',
        'event_id': event_id,
        'processed_at': json.dumps({'timestamp': 'now'})
    }

    # Store the result using the Idempotency-Key
    if idempotency_key:
        processed_events[idempotency_key] = response_data

    return jsonify(response_data), 200

if __name__ == '__main__':
    # Generate a random secret for demonstration
    app.config['HMAC_SECRET'] = secrets.token_hex(32)
    app.run(debug=True)